### PR TITLE
Add PESignerBuilder

### PR DESCRIPTION
**Pull request by micpe083**
Added a builder for PESigner to share the build logic.

Please let me know your feedback.

Cheers,
Mike

### CODE REVIEW

**Comment by ebourg** [comment]
Thank you for the contribution Michael, this will be useful for the Maven plugin too. Before merging could you just adjust the following points please:
- make PESignerBuilder package private (it's for internal use only)
- adjust PESignerBuilder code style (opening curly braces at the end of the lines)
- PESignerBuilder is missing a license header
- no need to put final everywhere
- PESignerBuilder.param() should probably throw an IllegalArgumentException for unknown parameters

There is also an issue with message logging. With the Ant task we use the Ant logger, with the CLI we use System.err/out, but in PESignerBuilder only System.err/out is used. So we need a thin logging interface to redirect the messages properly. We could reuse the one from jdeb:

https://github.com/tcurdt/jdeb/blob/master/src/main/java/org/vafer/jdeb/Console.java
https://github.com/tcurdt/jdeb/blob/master/src/main/java/org/vafer/jdeb/ant/TaskConsole.java
https://github.com/tcurdt/jdeb/blob/master/src/main/java/org/vafer/jdeb/maven/MojoConsole.java

**Comment by micpe083** [comment]
Hi Emmanuel, please see my feedback below:

- make PESignerBuilder package private (it's for internal use only)
 I think the builder should be public as the class it's building is also public. This will also give easy API access if someone needs it.

- adjust PESignerBuilder code style (opening curly braces at the end of the lines)
Done

- PESignerBuilder is missing a license header
Done

- no need to put final everywhere
Done

- PESignerBuilder.param() should probably throw an IllegalArgumentException for unknown parameters
Done

- Logger
Minimal logger added. Might be better to add slf4j or other api in future...

Cheers,
Mike

**Comment by ebourg** [comment]
Thank you for the modifications. Regarding the visibility of PESignerBuilder, the PESigner class already provides a builder like API with a fluent syntax and strongly typed methods. From a user point a view I think it's preferable to use this class. PESignerBuilder is useful to factorize the codes in the CLI/Ant task/Maven plugin, but I don't think it'll have other use cases.

**Comment by micpe083** [comment]
Looks like I forgot to mention this, but the reason I'm sending you this pull request is so I can use jsign without having to copy/paste the code in your cli/ant classes.

If I make the class private, then I can't use it and I'm back to square one again... :(

**Comment by ebourg** [comment]
Do you intend to use PESignerBuilder outside of PESignerTask and PESignerCLI?

**Comment by micpe083** [comment]
Yes

**Comment by ebourg** [comment]
Could you give me an example of how you'll use it and why it's better than using PESigner directly?

**Comment by micpe083** [comment]
I would use it like Option A below using the PEBuilder. Using PESigner directly (Option B) requires writing and maintaining much more code.

**Option A - create PESigner with PEBuilder:**

~~~~
PESignerBuilder builder = new PESignerBuilder(new PELogSysout(this));
builder.name(name);
builder.url(url);
builder.alg(algorithm);
builder.keystore(keystore);
builder.storepass(storepass);
builder.storetype(storetype);
builder.alias(alias);
builder.certfile(certfile);
builder.keyfile(keyfile);
builder.keypass(keypass);
builder.tsaurl(tsaurl);
builder.tsmode(tsmode);

PESigner signer = builder.build();
~~~~

**Option B - create PESigner without PEBuilder:**

~~~~
if (keystore != null && storetype == null) {
	// guess the type of the keystore from the extension of the file
	String filename = keystore.getName().toLowerCase();
	if (filename.endsWith(".p12") || filename.endsWith(".pfx")) {
		storetype = "PKCS12";
	} else {
		storetype = "JKS";
	}
}

PrivateKey privateKey;
Certificate[] chain;

// some exciting parameter validation...
if (keystore == null && keyfile == null && certfile == null) {
	throw new SignerException("keystore option, or keyfile and certfile options must be set");
}
if (keystore != null && (keyfile != null || certfile != null)) {
	throw new SignerException("keystore option can't be mixed with keyfile or certfile");
}

if (keystore != null) {
	// JKS or PKCS12 keystore
	KeyStore ks;
	try {
		ks = KeyStore.getInstance(storetype);
	} catch (KeyStoreException e) {
		throw new SignerException("keystore type '" + storetype + "' is not supported", e);
	}

	if (!keystore.exists()) {
		throw new SignerException("The keystore " + keystore + " couldn't be found");
	}
	FileInputStream in = null;
	try {
		in = new FileInputStream(keystore);
		ks.load(in, storepass != null ? storepass.toCharArray() : null);
	} catch (Exception e) {
		throw new SignerException("Unable to load the keystore " + keystore, e);
	} finally {
		try {
			if (in != null) {
				in.close();
			}
		} catch (IOException e) {
			// ignore
		}
	}

	if (alias == null) {
		throw new SignerException("alias option must be set");
	}

	try {
		chain = ks.getCertificateChain(alias);
	} catch (KeyStoreException e) {
		throw new SignerException(e.getMessage(), e);
	}
	if (chain == null) {
		throw new SignerException("No certificate found under the alias '" + alias + "' in the keystore " + keystore);
	}

	char[] password = keypass != null ? keypass.toCharArray() : storepass.toCharArray();

	try {
		privateKey = (PrivateKey) ks.getKey(alias, password);
	} catch (Exception e) {
		throw new SignerException("Failed to retrieve the private key from the keystore", e);
	}

} else {
	// separate private key and certificate files (PVK/SPC)
	if (keyfile == null) {
		throw new SignerException("keyfile option must be set");
	}
	if (!keyfile.exists()) {
		throw new SignerException("The keyfile " + keyfile + " couldn't be found");
	}
	if (certfile == null) {
		throw new SignerException("certfile option must be set");
	}
	if (!certfile.exists()) {
		throw new SignerException("The certfile " + certfile + " couldn't be found");
	}

	// load the certificate chain
	try {
		chain = loadCertificateChain(certfile);
	} catch (Exception e) {
		throw new SignerException("Failed to load the certificate from " + certfile, e);
	}

	// load the private key
	try {
		privateKey = PVK.parse(keyfile, keypass);
	} catch (Exception e) {
		throw new SignerException("Failed to load the private key from " + keyfile, e);
	}
}

if (alg != null && DigestAlgorithm.of(alg) == null) {
	throw new SignerException("The digest algorithm " + alg + " is not supported");
}

try {
	initializeProxy(proxyUrl, proxyUser, proxyPass);
} catch (Exception e) {
	throw new SignerException("Couldn't initialize proxy ", e);
}

PESigner signer = new PESigner(chain, privateKey)
		.withLog(log)
		.withProgramName(name)
		.withProgramURL(url)
		.withDigestAlgorithm(DigestAlgorithm.of(alg))
		.withTimestamping(tsaurl != null || tsmode != null)
		.withTimestampingMode(tsmode != null ? TimestampingMode.of(tsmode) : TimestampingMode.AUTHENTICODE)
		.withTimestampingAutority(tsaurl);

@SuppressWarnings("unchecked")
private Certificate[] loadCertificateChain(File file) throws IOException, CertificateException {
	FileInputStream in = null;
	try {
		in = new FileInputStream(file);
		CertificateFactory certificateFactory = CertificateFactory.getInstance("X.509");
		Collection<Certificate> certificates = (Collection<Certificate>) certificateFactory.generateCertificates(in);
		return certificates.toArray(new Certificate[certificates.size()]);
	} finally {
		try {
			if (in != null) {
				in.close();
			}
		} catch (IOException e) {
			// ignore
		}
	}
}

private void initializeProxy(String proxyUrl, final String proxyUser, final String proxyPassword) throws MalformedURLException {
	// Do nothing if there is no proxy url.
	if (proxyUrl != null && proxyUrl.trim().length() > 0) {
		if (!proxyUrl.trim().startsWith("http")) {
			proxyUrl = "http://" + proxyUrl.trim();
		}
		final URL url = new URL(proxyUrl);
		final int port = url.getPort() < 0 ? 80 : url.getPort();

		ProxySelector.setDefault(new ProxySelector() {
			private List<Proxy> proxies = Collections.singletonList(new Proxy(Proxy.Type.HTTP, new InetSocketAddress(url.getHost(), port)));

			public List<Proxy> select(URI uri) {
				return proxies;
			}

			public void connectFailed(URI uri, SocketAddress sa, IOException ioe) {
			}
		});

		if (proxyUser != null && proxyUser.length() > 0 && proxyPassword != null) {
			Authenticator.setDefault(new Authenticator() {
				protected PasswordAuthentication getPasswordAuthentication() {
					return new PasswordAuthentication(proxyUser, proxyPassword.toCharArray());
				}
			});
		}
	}
}
~~~~

**Comment by ebourg** [comment]
Ok understood, thank you for the example. Maybe more validation logic should be moved into PESigner, I'll think about it.

File: src/main/java/net/jsign/PESigner.java [hunk 1]
```diff
@@ -163,17 +176,56 @@ public void sign(PEFile file) throws Exception {
...
+
         // compute the signature
         CertificateTableEntry entry = createCertificateTableEntry(file);
-        
+
         file.writeDataDirectory(DataDirectoryType.CERTIFICATE_TABLE, entry.toBytes());
         file.close();
     }
 
+    public void sign(String file) throws SignerException {
```
Comment (netmackan) [inline src/main/java/net/jsign/PESigner.java]: This is just a shortcut for new File(String) but makes the API more unclear as PESigner.sign(String) does not tell the user that the String should be a file name. Better only have the strongly typed PESigner.sign(File).

**Comment by netmackan** [comment]
I will continue to use the PESigner API which suites me very well and as I normally do not load the KeyStore instances from files and use custom Timestamper implementations and so on. That's why I prefer not mixing in to much high level logic in the PESigner class. That being said, having a simpler builder class in addition to that, as suggested here, could probably be useful to some.

**Comment by ebourg** [comment]
I rebased the PR and made some additional modifications, thank you Michael.

**Pull request closed without merging**